src/som/interpreter/bc/interpreter.py

```diff
@@ -702,38 +704,43 @@ def get_self(frame, ctx_level):
 
 @elidable_promote("all")
 def _lookup(layout, selector, method, bytecode_index):
-    # First try of inline cache
-    cached_layout1 = method.get_inline_cache_layout(bytecode_index)
-    if cached_layout1 is layout:
-        invokable = method.get_inline_cache_invokable(bytecode_index)
-    elif cached_layout1 is None:
-        invokable = layout.lookup_invokable(selector)
-        method.set_inline_cache(bytecode_index, layout, invokable)
-    else:
-        # second try
-        # the bytecode index after the send is used by the selector constant,
-        # and can be used safely as another cache item
-        cached_layout2 = method.get_inline_cache_layout(bytecode_index + 1)
-        if cached_layout2 == layout:
-            invokable = method.get_inline_cache_invokable(bytecode_index + 1)
-        else:
-            invokable = layout.lookup_invokable(selector)
-            if cached_layout2 is None:
-                method.set_inline_cache(bytecode_index + 1, layout, invokable)
-    return invokable
+    cache = first = method.get_inline_cache(bytecode_index)
+    while cache is not None:
+        if cache.expected_layout is layout:
+            return cache.get_cached_method()
+        cache = cache.next_entry
+
+    invoke = layout.lookup_invokable(selector)
+
+    cache_size = 0
+    cache = first
+    while cache is not None:
+        cache_size += 1
+        cache = cache.next_entry
+
+    if INLINE_CACHE_SIZE >= cache_size:
+        method.set_inline_cache(bytecode_index, CachedDispatchNode(rcvr_class=layout, method=invoke, next_entry=first))
+
+    return invoke
 
 
 def _update_object_and_invalidate_old_caches(obj, method, bytecode_index, universe):
     obj.update_layout_to_match_class()
     obj.get_object_layout(universe)
 
-    cached_layout1 = method.get_inline_cache_layout(bytecode_index)
-    if cached_layout1 is not None and not cached_layout1.is_latest:
-        method.set_inline_cache(bytecode_index, None, None)
+    prev = None
```

Comment:
Can we avoid the code duplication here?
Not sure it's practical since the AST interpreter works on a node here though. hmm...

Reply:
The code duplication with `_AbstractGenericMessageNode._get_cache_size_and_drop_old_entries`, since that's where I got the logic from? Or do you mean the redundant linked list walks with 
```
cache = method.get_inline_cache(bytecode_index)
while cache is not None:
    ...
    cache = cache.next_entry
```
like in `_lookup` ?

Reply:
The code duplication between the AST and BC interpreters, so I guess `_AbstractGenericMessageNode._get_cache_size_and_drop_old_entries`